Source/Core/Core/NetPlayServer.cpp

```diff
@@ -733,6 +756,57 @@ bool NetPlayServer::StartGame()
   return true;
 }
 
+void NetPlayServer::UpdateBuffer()
+{
+  std::map<PlayerId, u32> players_selected_pings;
+
+  for (auto& player : m_players)
+  {
+    if (player.second.last_pings.size() < AUTO_BUFFER_PING_WINDOW_SEC)
+    {
+      return;  // wait for window to fill
+    }
+
+    std::deque<u32> player_last_pings = player.second.last_pings;
+    sort(player_last_pings.begin(), player_last_pings.end());
+    size_t index =
+        static_cast<size_t>(round((player_last_pings.size() - 1) * m_auto_buffer_sensibility));
+    index = std::min(index, player_last_pings.size() - 1);
+
+    players_selected_pings[player.second.pid] = player_last_pings[index];
+  }
+
+  u32 max_player_pair_ping = 0;
+
+  // get the max player pair ping to account for client -> server -> client
+  // when they are more than two players
+  for (auto& p1 : m_players)
+  {
+    for (auto& p2 : m_players)
+    {
+      if (p1.second == p2.second)
+        continue;
+
+      u32 p1_ping = players_selected_pings[p1.second.pid];
+      u32 p2_ping = players_selected_pings[p2.second.pid];
+
+      max_player_pair_ping = std::max(max_player_pair_ping, p1_ping + p2_ping);
+    }
+  }
+
+  u32 poll_rate = SerialInterface::GetPollXFrame() > 0 ? SerialInterface::GetPollXFrame() :
+                                                         DEFAULT_PAD_POLL_RATE;
+
+  u32 refresh_rate = VideoInterface::GetTargetRefreshRate() > 0 ?
```

Comment:
Why do you depend on VI scanout configuration? I'd suggest to not use any emulation state here but maybe GCpad configuration.